index/segment/fst/pilosa_postings_codec.go

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fst
+
+import (
+	"bytes"
+
+	"github.com/m3db/m3ninx/postings"
+	pilosa "github.com/pilosa/pilosa/roaring"
+)
+
+type pilosaPostingsEncoder struct {
+	scratchBuffer bytes.Buffer
+}
+
+// TODO(prateek): take an initial size parameter
+func newPilosaPostingsEncoder() *pilosaPostingsEncoder {
+	return &pilosaPostingsEncoder{}
+}
+
+func (p *pilosaPostingsEncoder) Reset() {
+	p.scratchBuffer.Reset()
+}
+
+func (p *pilosaPostingsEncoder) Serialize(pl postings.List) ([]byte, error) {
+	p.scratchBuffer.Reset()
+
+	pilosaBitmap, err := toPilosa(pl)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = pilosaBitmap.WriteTo(&p.scratchBuffer)
+	if err != nil {
+		return nil, err
+	}
+
+	return p.scratchBuffer.Bytes(), nil
+}
+
+func toPilosa(pl postings.List) (*pilosa.Bitmap, error) {
+	bitmap := pilosa.NewBitmap()
+	iter := pl.Iterator()
+
+	for iter.Next() {
+		_, err := bitmap.Add(uint64(iter.Current()))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if err := iter.Err(); err != nil {
+		return nil, err
+	}
+
+	return bitmap, nil
+}
+
+func fromPilosa(data []byte, pool postings.Pool) (postings.List, error) {
+	b := pilosa.NewBitmap()
+	if err := b.UnmarshalBinary(data); err != nil {
+		return nil, err
+	}
+
+	pl := pool.Get()
+	iter := newIter(b.Iterator())
+	return pl, pl.AddIterator(iter)
+}
+
+// NB: need to do this to find a path into our postings list which doesn't require every
+// insert to grab a lock. we need to make a non thread-safe version of our api.
+
+type pilosaBitmapIterator struct {
+	iter    *pilosa.Iterator
+	current uint64
+	hasNext bool
+}
+
+func newIter(iter *pilosa.Iterator) *pilosaBitmapIterator {
+	return &pilosaBitmapIterator{
+		iter: iter,
+	}
+}
+
+func (p *pilosaBitmapIterator) Next() bool {
+	v, eof := p.iter.Next()
```

Comment:
Does this give EOF just once?  If so should the EOF be cached, perhaps `if !p.hasNext { return false }` early from this method?

Reply:
fair will do